ReactiveCocoa/Swift/Signal.swift

```diff
@@ -410,16 +410,126 @@ public func takeUntil<T, E>(trigger: Signal<(), NoError>)(signal: Signal<T, E>)
+	let liftingDisposable = scannedSignal.observe(liftingProducerSink)
+	
+	var outputSignal: Signal<U, E>!
+	let producerOfLastValue = SignalProducer(value: initial)
+		|> concat(liftingProducerForScannedSignal)
+		|> takeLast(1)
+	producerOfLastValue.startWithSignal { startedOutputSignal, disposable in
+		disposable.addDisposable(liftingDisposable)
+		outputSignal = startedOutputSignal
+		// No guarantee here that outputSignal won't already have completed by
+		// the time the caller observes it.
+	}
+	return outputSignal
+}
+
+/// Aggregates `signal`'s values into a single combined value. When `signal` emits
+/// its first value, `combine` is invoked with `initial` as the first argument and
+/// that emitted value as the second argument. The result is emitted from the
+/// signal returned from `reduce`. That result is then passed to `combine` as the
+/// first argument when the next value is emitted, and so on.
+public func scan<T, U, E>(initial: U, combine: (U, T) -> U)(signal: Signal<T, E>) -> Signal<U, E> {
+	return Signal { observer in
+		var accumulator = initial
+		return signal.observe(Signal.Observer { event in
+			observer.put(event.map { value in
+				accumulator = combine(accumulator, value)
+				return accumulator
+			})
+		})
+	}
+}
+
+/// Forwards only those values from `signal` which are not duplicates of the
+/// immedately preceding value. The first value is always forwarded.
+public func skipRepeats<T: Equatable, E>(signal: Signal<T, E>) -> Signal<T, E> {
+	return signal |> skipRepeats { $0 == $1 }
+}
+
+/// Forwards only those values from `signal` which do not pass `isRepeat` with
+/// respect to the previous value. The first value is always forwarded.
+public func skipRepeats<T, E>(isRepeat: (T, T) -> Bool)(signal: Signal<T, E>) -> Signal<T, E> {
+	return signal
+		|> map { Optional($0) }
+		|> combinePrevious(nil)
+		|> filter {
+			switch $0 {
+			case let (.Some(a), .Some(b)) where isRepeat(a, b):
+				return false
+			default:
+				return true
+			}
+		}
+		|> map { $0.1! }
+}
+
+/// Does not forward any values from `signal` until `predicate` returns false,
+/// at which point the returned signal behaves exactly like `signal`.
+public func skipWhile<T, E>(predicate: T -> Bool)(signal: Signal<T, E>) -> Signal<T, E> {
+	return Signal { observer in
+		var shouldSkip = true
+		return signal.observe(next: { value in
+			shouldSkip = shouldSkip && predicate(value)
+			if !shouldSkip {
+				sendNext(observer, value)
+			}
+		}, error: { error in
+			sendError(observer, error)
+		}, completed: {
+			sendCompleted(observer)
+		})
+	}
+}
+
+/// Waits until `signal` completes and then forwards the final `count` values
+/// on the returned signal.
+public func takeLast<T,E>(count: Int)(signal: Signal<T,E>) -> Signal<T,E> {
+	return Signal { observer in
+		var buffer = [T]()
+		buffer.reserveCapacity(count)
+
+		return signal.observe(next: { value in
+			buffer.append(value)
+
+			while buffer.count > count {
```

Comment:
Doesn't this mean we've exceeded the reserved capacity? Not that that's an error—but it seems to defeat the potential optimization.

Should we remove and _then_ append instead?